executor/aggfuncs/func_max_min.go

```diff
@@ -0,0 +1,262 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aggfuncs
+
+import (
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+)
+
+type partialResult4MaxMinInt = int64
+type partialResult4MaxMinUint = uint64
+type partialResult4MaxMinDecimal = *types.MyDecimal
+type partialResult4MaxMinFloat32 = float32
+type partialResult4MaxMinFloat64 = float64
+
+// Todo
+type partialResult4MaxMinString string
+type partialResult4MaxMinDatetime types.Time
+type partialResult4MaxMinTimestamp types.Time
+type partialResult4MaxMinDuration types.Duration
+
+type baseMaxMinAggFunc struct {
+	baseAggFunc
+
+	isMax bool
+	// executed is used to indicates:
+	// 1. whether the partial result is the initialization value which should not be compared during evaluation;
+	// 2. whether all the values of arg are all null, if so, we should return null as the default value for MAX/MIN.
+	executed bool
```

Comment:
this should be in the field of partial results, and should be reset for every group